xarray/core/coordinates.py

```diff
@@ -215,7 +215,15 @@ def __getitem__(self, key: Hashable) -> "DataArray":
 
     def to_dataset(self) -> "Dataset":
         """Convert these coordinates into a new Dataset"""
-        return self._data._copy_listed(self._names)
+
+        def key(name):
+            try:
+                return list(self._data._variables.keys()).index(name)
+            except ValueError:
+                return len(self._data._variables)
+
+        names = sorted(self._names, key=key)
```

Comment:
It would be slightly better to iterate over `_variables` only once. That would run in O(N) time rather than O(N**2).

e.g.,

```python
names = [k for k in self._data._variables if k in self._names]
```

Reply:
thanks, that's better